browser/brave_shields/reduce_language_navigation_throttle.cc

```diff
@@ -0,0 +1,114 @@
+#include "content/public/browser/navigation_handle.h"
+#include "content/public/browser/web_contents.h"
+#include "net/http/http_util.h"
+#include "third_party/blink/renderer/platform/language.h"
+#include "third_party/blink/renderer/platform/wtf/text/wtf_string.h"
+
+namespace brave_shields {
+
+// static
+std::unique_ptr<ReduceLanguageNavigationThrottle>
+ReduceLanguageNavigationThrottle::MaybeCreateThrottleFor(
+    content::NavigationHandle* navigation_handle,
+    HostContentSettingsMap* content_settings) {
+  if (!base::FeatureList::IsEnabled(
+          brave_shields::features::kBraveReduceLanguage))
+    return nullptr;
+  return std::make_unique<ReduceLanguageNavigationThrottle>(navigation_handle,
+                                                            content_settings);
+}
+
+ReduceLanguageNavigationThrottle::ReduceLanguageNavigationThrottle(
+    content::NavigationHandle* navigation_handle,
+    HostContentSettingsMap* content_settings)
+    : content::NavigationThrottle(navigation_handle),
+      content_settings_(content_settings) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+}
+
+ReduceLanguageNavigationThrottle::~ReduceLanguageNavigationThrottle() {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+}
+
+content::NavigationThrottle::ThrottleCheckResult
+ReduceLanguageNavigationThrottle::WillStartRequest() {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  UpdateHeaders();
+  return content::NavigationThrottle::PROCEED;
+}
+
+content::NavigationThrottle::ThrottleCheckResult
+ReduceLanguageNavigationThrottle::WillRedirectRequest() {
+  return WillStartRequest();
+}
+
+void ReduceLanguageNavigationThrottle::UpdateHeaders() {
+  content::NavigationHandle* handle = navigation_handle();
+  GURL visible_url = handle->GetWebContents()->GetVisibleURL();
+  content::BrowserContext* context =
+      handle->GetWebContents()->GetBrowserContext();
+  PrefService* pref_service = user_prefs::UserPrefs::Get(context);
+
+  if (!brave_shields::ShouldDoReduceLanguage(content_settings_, visible_url,
+                                             pref_service))
+    return;
+
+  ControlType fingerprinting_control_type =
+      brave_shields::GetFingerprintingControlType(content_settings_,
+                                                  visible_url);
+
+  // If fingerprint blocking is maximum, set Accept-Language header to
+  // static value regardless of other preferences.
+  if (fingerprinting_control_type == ControlType::BLOCK) {
+    handle->SetRequestHeader(net::HttpRequestHeaders::kAcceptLanguage,
+                             "en-US,en");
+    return;
+  }
+
+  // If fingerprint blocking is default, compute Accept-Language header
+  // based on user preferences.
+  std::string languages =
+      pref_service->Get(language::prefs::kAcceptLanguages)->GetString();
+  std::string first_language = language::GetFirstLanguage(languages);
+  std::vector<std::string> q_values = {";q=0.5", ";q=0.6", ";q=0.7",
```

Comment:
what are these q values? Can you add comments explaining?

Reply:
and also link to the spec if relevant

Reply:
added comment in https://github.com/brave/brave-core/pull/12234/commits/fc7510cffc3499d49b9c85ac7e3536e4cc0b5d01